torchvision/csrc/models/vgg.cpp

```diff
@@ -0,0 +1,116 @@
+#include "vgg.h"
+
+#include <map>
+#include "modelsimpl.h"
+
+namespace vision {
+namespace models {
+torch::nn::Sequential makeLayers(
+    const std::vector<int>& cfg,
+    bool batch_norm = false) {
+  torch::nn::Sequential seq;
+  auto channels = 3;
+
+  for (const auto& V : cfg) {
+    if (V <= -1)
+      seq->push_back(modelsimpl::MaxPool2D(2, 2));
+    else {
+      seq->push_back(torch::nn::Conv2d(
+          torch::nn::Conv2dOptions(channels, V, 3).padding(1)));
+
+      if (batch_norm)
+        seq->push_back(torch::nn::BatchNorm(V));
+      seq->push_back(modelsimpl::Relu(true));
+
+      channels = V;
+    }
+  }
+
+  return seq;
+}
+
+void VGGImpl::_initialize_weights() {
+  for (auto& module : modules(false)) {
+    if (torch::nn::Conv2dImpl* M =
+            dynamic_cast<torch::nn::Conv2dImpl*>(module.get())) {
+      torch::nn::init::xavier_normal_(M->weight); // TODO kaiming
+      torch::nn::init::constant_(M->bias, 0);
+    } else if (
+        torch::nn::BatchNormImpl* M =
+            dynamic_cast<torch::nn::BatchNormImpl*>(module.get())) {
+      torch::nn::init::constant_(M->weight, 1);
+      torch::nn::init::constant_(M->bias, 0);
+    } else if (
+        torch::nn::LinearImpl* M =
+            dynamic_cast<torch::nn::LinearImpl*>(module.get())) {
+      torch::nn::init::normal_(M->weight, 0, 0.01);
+      torch::nn::init::constant_(M->bias, 0);
+    }
+  }
+}
+
+VGGImpl::VGGImpl(
+    torch::nn::Sequential features,
+    int64_t num_classes,
+    bool initialize_weights) {
+  // clang-format off
+    classifier = torch::nn::Sequential(
+		torch::nn::Linear(512 * 7 * 7, 4096),
+		modelsimpl::Relu(true),
+		torch::nn::Dropout(),
+		torch::nn::Linear(4096, 4096),
+		modelsimpl::Relu(true),
+		torch::nn::Dropout(),
+		torch::nn::Linear(4096, num_classes));
+  // clang-format on
+
+  this->features = features;
+
+  register_module("features", this->features);
+  register_module("classifier", classifier);
+
+  if (initialize_weights)
+    _initialize_weights();
+}
+
+torch::Tensor VGGImpl::forward(torch::Tensor x) {
+  x = features->forward(x);
+  x = x.view({x.size(0), -1});
+  x = classifier->forward(x);
+  return x;
+}
+
+// clang-format off
+static std::map<char, std::vector<int>> cfg = {
```

Comment:
I recommend using `unordered_map` (hashtable) instead of `map` (binary tree)

Reply:
I did think of this but then I thought that because of the size of the map it doesn't matter here.